### PR TITLE
PAINTROID-634 : Add Eraser tool

### PR DESCRIPTION
**Pull request by bakicelebi**
[PAINTROID-634](https://jira.catrob.at/browse/PAINTROID-634)
[PAINTROID-533](https://jira.catrob.at/browse/PAINTROID-533)
[PAINTROID-646](https://jira.catrob.at/browse/PAINTROID-646)

## New Features and Enhancements

- **Eraser Tool**: Added an eraser tool that functions similarly to the brush tool but utilizes a different BlendMode.
- **Constant Identifier Names Analysis**: Added a new analysis option [constant_identifier_names](https://dart.dev/tools/linter-rules/constant_identifier_names) which is set to `false` to stay consistent with Paintroid-Kotlin.
- **Clipboard Icon**: Copied the clipboard icon asset from Paintroid-Kotlin.
- **BlendMode and StrokeJoin Serialization**: Expanded the paint serializer to include support for BlendMode and strokeJoin properties.
- **ToolType Differentiation**: Assigned a ToolType to each tool, enhancing tool management and identification.
- **ModalBottomSheet for Tools**: Implemented a ModalBottomSheet that displays a list of available tools.
- **Navigation Enhancement**: Added the BottomNavBar functionality to highlight the active tool.
- **Toast Package Update**: Replaced the toast package with [this package](https://pub.dev/packages/toast) for better toast notifications.
- **Integration Test Improvements**: Changed integration testing through the addition of utility functions and the reorganization of existing tests.

## Refactorings and Bug Fixes

- **Component Renaming**: Renamed the component "BottomControlNavigationBar" to "BottomNavBar" for consistency and clarity.
- **Background and PaintingLayer Refactoring**: Refactored the BackgroundLayer and PaintingLayer widgets for improved readability.
- **PaintSerializer Optimization**: Optimized the PaintSerializer class.
- **Image Saving Improvement**: Reworked the image-saving mechanism to fix JPG transparency issues that arose when using the eraser tool.
- **Bug Fix**: Addressed a bug where strokes were saved with incorrect stroke widths in projects/images.
- **API Permission Handling**: Fixed permission issues for API versions greater than 28.

## Checklist
- [x] Include the name of the Jira ticket in the PR’s title
- [x] Include a summary of the changes plus the relevant context
- [x] Choose the proper base branch (*develop*)
- [x] Confirm that the changes follow the project’s coding guidelines
- [x] Verify that the changes generate no compiler or linter warnings
- [x] Perform a self-review of the changes
- [x] Verify to commit no other files than the intentionally changed ones
- [x] Include reasonable and readable tests verifying the added or changed behavior
- [x] Confirm that new and existing unit tests pass locally
- [x] Check that the commits’ message style matches the [project’s guideline](https://github.com/Catrobat/Catroid/wiki/Commit-Message-Guidelines)
- [x] Stick to the project’s gitflow workflow
- [x] Verify that your changes do not have any conflicts with the base branch
- [x] After the PR, verify that all CI checks have passed
- [x] Post a message in the *#paintroid* [Slack channel](https://catrobat.slack.com) and ask for a code reviewer

### CODE REVIEW

File: lib/io/src/serialization/proto/schema/graphic/paint.proto [hunk 1]
```diff
@@ -14,4 +14,16 @@ message SerializablePaint {
...
+  enum BlendMode {
+    SCR_OVER = 0;
+    CLEAR = 1;
+  }
+  BlendMode blendMode = 5;
+
+  enum StrokeJoin {
+    SJ_MITER = 0;
+    SJ_ROUND = 1;
+    SJ_BEVEL = 2;
```
Comment (juliajulie95) [inline lib/io/src/serialization/proto/schema/graphic/paint.proto]: I think the names are difficult to read, can we name them more clearly?

File: lib/io/src/service/permission_service.dart [hunk 2]
```diff
@@ -53,7 +72,11 @@ class PermissionService with LoggableMixin implements IPermissionService {
     if (Platform.isIOS) {
       permission = Permission.photosAddOnly;
     } else if (Platform.isAndroid) {
-      permission = Permission.storage;
+      if (await _isAndroidVersionGreaterOrEqualTo(33)) {
```
Comment (juliajulie95) [inline lib/io/src/service/permission_service.dart:75]: Did you test it with older api versions?
I think this was related to scoped storage which was introduced with android 29 I think

File: lib/ui/drawing_space/bottom_nav_bar.dart [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+  }
+
+  @override
+  Widget build(BuildContext context) {
+    final localizations = AppLocalizations.of(context);
+    return NavigationBarTheme(
+      data: NavigationBarThemeData(
+        indicatorColor: Colors.transparent,
+        labelTextStyle: MaterialStateProperty.all(
+          TextStyle(color: Theme.of(context).colorScheme.onSurface),
```
Comment (juliajulie95) [inline lib/ui/drawing_space/bottom_nav_bar.dart]: The ThemeData should go to the styles file

File: lib/ui/drawing_space/tools_bottom_sheet.dart [hunk 4]
```diff
@@ -0,0 +1,28 @@
...
+
+class ToolsBottomSheet extends StatelessWidget {
+  const ToolsBottomSheet({
+    super.key,
+  });
+
+  @override
+  Widget build(BuildContext context) {
+    const tools = ToolData.allToolsData;
+    return Column(
```
Comment (juliajulie95) [inline lib/ui/drawing_space/tools_bottom_sheet.dart]: Did you try this?
https://api.flutter.dev/flutter/widgets/GridView/GridView.count.html

File: lib/ui/shared/icon_button_with_label.dart [hunk 5]
```diff
@@ -0,0 +1,54 @@
+import 'package:flutter/material.dart';
+import 'package:flutter_riverpod/flutter_riverpod.dart';
+import 'package:flutter_svg/flutter_svg.dart';
+import 'package:paintroid/tool/tool.dart';
+import 'package:toast/toast.dart';
+
+class IconButtonWithLabel extends StatelessWidget {
```
Comment (juliajulie95) [inline lib/ui/shared/icon_button_with_label.dart:3]: We are mixing too much UI with logic in here.
Could we make this widget 100% generic and hand over the function as a callback and remove the tooldata?

If you do not like that solution, renaming it to something more specific is fine for me too.
But now it seems like a very generic widget, but it changes the tool

The 100% perfect solution would be a very generic IconButtonWithLabel that gets icon, label & callback
And then a ToolButton which uses the IconButtonWithLabel and implements the tool switching

File: lib/workspace/src/usecase/render_image_for_export.dart [hunk 6]
```diff
@@ -25,28 +25,51 @@ class RenderImageForExport {
...
+        _graphicFactory.createCanvasWithRecorder(commandsRecorder);
+    commandsCanvas.clipRect(scaledRect, doAntiAlias: false);
+    _commandManager.executeAllCommands(commandsCanvas);
+
+    final combinedRecorder = _graphicFactory.createPictureRecorder();
+    final combinedCanvas =
+        _graphicFactory.createCanvasWithRecorder(combinedRecorder);
+
+    final bgPicture = backgroundRecorder.endRecording();
+    final bgImage = await bgPicture.toImage(
```
Comment (juliajulie95) [inline lib/workspace/src/usecase/render_image_for_export.dart]: I would write the whole name like you did with the others
backgroundPicture
backgroundImage
Also those 2 are synonyms, maybe specify what makes them different in the name

**Review by juliajulie95 — CHANGES_REQUESTED** [review]
Added some comments, looks good to me though!

File: lib/io/src/serialization/proto/schema/graphic/paint.proto (hunk 1 above)
Comment (bakicelebi) [reply lib/io/src/serialization/proto/schema/graphic/paint.proto]: Added their respective enum names before each enum element. Should also cause fewer problems in the future when we add more enums.

File: lib/ui/drawing_space/bottom_nav_bar.dart (hunk 3 above)
Comment (bakicelebi) [reply lib/ui/drawing_space/bottom_nav_bar.dart]: Added WidgetThemes abstract class, alse renamed ThemeText to TextThemes.

File: lib/ui/drawing_space/tools_bottom_sheet.dart (hunk 4 above)
Comment (bakicelebi) [reply lib/ui/drawing_space/tools_bottom_sheet.dart]: Refactored to Gridview.count, better for readability. Thanks

File: lib/ui/shared/icon_button_with_label.dart (hunk 5 above)
Comment (bakicelebi) [reply lib/ui/shared/icon_button_with_label.dart:3]: Create a new ToolButton widget, moved toast to provider,

File: lib/workspace/src/usecase/render_image_for_export.dart (hunk 6 above)
Comment (bakicelebi) [reply lib/workspace/src/usecase/render_image_for_export.dart]: Refactored the function, it should be more readable now

File: lib/io/src/service/permission_service.dart (hunk 2 above)
Comment (bakicelebi) [reply lib/io/src/service/permission_service.dart:75]: I tested it on API 34,33,28 and 24 and all of  them are working fine.

**Review by juliajulie95 — APPROVED** [review]

**Pull request merged**